### PR TITLE
Asset conversion nits

### PR DESCRIPTION
**Pull request by bkontur**
https://github.com/paritytech/substrate/pull/14572

PR:
- fixes benchmarks for pallet_asset_conversion - `create_pool` error
- small refactor to reuse conversion stuff for `assetId` vs `MultiLocation`
- Allow `set_storage` for `AllowMultiAssetPools` / `LiquidityWithdrawalFee`
- small nits

TODOs
- [x] native asset - migration or storage set for westend - https://github.com/paritytech/cumulus/pull/2860#pullrequestreview-1528188960
- [x] set correct weight instead `pallet_asset_conversion::Config -> type WeightInfo = ();` when benchmarks start to work
- [x] (can be separate PR) fix benchmarks - https://github.com/paritytech/cumulus/issues/2874
- [ ] ~~change AssetConversionApi to use `VersionedMultiLocation`~~ continues here https://github.com/paritytech/cumulus/issues/2915

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3178765

**Review by joepetrowski — APPROVED** [review]
We already have some pools with `parents: 0`. I don't mind nuking them on Westend, but could you either provide a migration or a `set_storage` call to sudo?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3186114

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3191146

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3191733

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3191778

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3193718

**Comment by bkontur** [comment]
bot fmt

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3194072

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3194137

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3202826

**Comment by bkontur** [comment]
bot bench $ pallet dev pallet_asset_conversion

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3204264

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3211402

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3215877

File: parachains/runtimes/assets/common/src/local_and_foreign_assets.rs [hunk 1]
```diff
@@ -13,92 +13,82 @@
...
-	fn try_convert(asset_id: &Box<MultiLocation>) -> Result<MultiLocation, ()> {
-		let mut asset_id = asset_id.clone();
-		asset_id.simplify(&ParachainLocation::get());
+	fn try_convert(
+		asset_id: &Box<MultiLocation>,
+	) -> MultiAssetIdConversionResult<Box<MultiLocation>, MultiLocation> {
 		if Self::is_native(&asset_id) {
 			// Otherwise it will try and touch the asset to create an account.
-			return Err(())
+			return MultiAssetIdConversionResult::Native
```
Comment (joepetrowski) [inline parachains/runtimes/assets/common/src/local_and_foreign_assets.rs:56]: What about this comment?

File: parachains/runtimes/assets/asset-hub-westend/src/lib.rs [hunk 2]
```diff
@@ -1360,3 +1376,113 @@ cumulus_pallet_parachain_system::register_validate_block! {
...
+
+			let mut reads = 1;
+			let mut writes = 0;
+
+			// migrate pools with invalid native asset
+			let pools = pallet_asset_conversion::Pools::<T>::iter().collect::<Vec<_>>();
+			for (old_pool_id, pool_info) in pools {
+				let old_pool_account =
+					pallet_asset_conversion::Pallet::<T>::get_pool_account(&old_pool_id);
+				reads += 1;
```
Comment (joepetrowski) [inline parachains/runtimes/assets/asset-hub-westend/src/lib.rs:1425]: Probably fine, but paranoia.
```suggestion
				reads.saturating_accrue(1);
```

File: parachains/runtimes/assets/asset-hub-westend/src/lib.rs [hunk 3]
```diff
@@ -1360,3 +1376,113 @@ cumulus_pallet_parachain_system::register_validate_block! {
...
+					old_pool_id.1.clone(),
+				);
+				let new_pool_account =
+					pallet_asset_conversion::Pallet::<T>::get_pool_account(&new_pool_id);
+				frame_system::Pallet::<T>::inc_providers(&new_pool_account);
+				reads += 2;
+				writes += 1;
+
+				// move currency
+				let _ = T::Currency::transfer(
```
Comment (joepetrowski) [inline parachains/runtimes/assets/asset-hub-westend/src/lib.rs:1444]: `transfer_all` simpler?

**Review by joepetrowski — COMMENTED** [review]
Looks OK, @jsidorenko can check for the migration especially?

File: parachains/runtimes/assets/asset-hub-westend/src/lib.rs (hunk 2 above)
Comment (bkontur) [reply parachains/runtimes/assets/asset-hub-westend/src/lib.rs]: ok, thx, fixed everywhere

File: parachains/runtimes/assets/common/src/local_and_foreign_assets.rs (hunk 1 above)
Comment (bkontur) [reply parachains/runtimes/assets/common/src/local_and_foreign_assets.rs]: removed

File: parachains/runtimes/assets/asset-hub-westend/src/lib.rs (hunk 3 above)
Comment (bkontur) [reply parachains/runtimes/assets/asset-hub-westend/src/lib.rs]: ok, thx, changed

**Comment by bkontur** [comment]
> Looks OK, @jsidorenko can check for the migration especially?

for that migration, I have a working test:
```
pub use asset_hub_westend_runtime::{AssetConversion, PoolAssets};

#[test]
fn test_migration() {
	use frame_support::traits::{
		fungible::Inspect, fungibles::roles::Inspect as InspectRoles, OnRuntimeUpgrade,
	};

	let account = AccountId::from(ALICE);
	let existential_deposit = ExistentialDeposit::get();

	ExtBuilder::<Runtime>::default()
		.with_collators(vec![AccountId::from(ALICE)])
		.with_session_keys(vec![(
			AccountId::from(ALICE),
			AccountId::from(ALICE),
			SessionKeys { aura: AuraId::from(sp_core::sr25519::Public::from_raw(ALICE)) },
		)])
		.with_balances(vec![
			(account.clone(), existential_deposit * AssetDeposit::get() * 100),
		])
		.build()
		.execute_with(|| {
			const ASSET_ID: u32 = 1;
			let asset_native = Box::new(MultiLocation::new(0, Here));
			let asset_one = Box::new(MultiLocation {
				parents: 0,
				interior: X2(PalletInstance(50), GeneralIndex(ASSET_ID.into())),
			});
			assert_ok!(Assets::create(
				RuntimeHelper::<Runtime>::origin_of(account.clone()),
				ASSET_ID.into(),
				account.clone().into(),
				1000,
			));
			assert_ok!(Assets::mint(
				RuntimeHelper::<Runtime>::origin_of(account.clone()),
				ASSET_ID.into(),
				account.clone().into(),
				3_000_000_000_000,
			));
			assert_ok!(AssetConversion::create_pool(
				RuntimeHelper::<Runtime>::origin_of(account.clone()),
				asset_native.clone(),
				asset_one.clone(),
			));
			assert_ok!(AssetConversion::add_liquidity(
				RuntimeHelper::<Runtime>::origin_of(account.clone()),
				asset_native.clone(),
				asset_one.clone(),
				1_000_000_000_000,
				2_000_000_000_000,
				0,
				0,
				account.into()
			));

			let lp_token = PoolAssets::asset_ids().next().unwrap();

			let old_pool_id = AssetConversion::get_pool_id(asset_native.clone(), asset_one.clone());
			let old_pool_account = AssetConversion::get_pool_account(&old_pool_id);
			let new_pool_id = AssetConversion::get_pool_id(Box::new(WestendLocation::get()), asset_one.clone());
			let new_pool_account = AssetConversion::get_pool_account(&new_pool_id);

			assert_ne!(0, Balances::balance(&old_pool_account));
			assert_eq!(0, Balances::balance(&new_pool_account));
			assert_ne!(0, Assets::balance(ASSET_ID, old_pool_account.clone()));
			assert_eq!(0, Assets::balance(ASSET_ID, new_pool_account.clone()));
			assert_ne!(0, PoolAssets::balance(lp_token, old_pool_account.clone()));
			assert_eq!(0, PoolAssets::balance(lp_token, new_pool_account.clone()));
			assert_eq!(Some(old_pool_account.clone()), PoolAssets::owner(lp_token));
			println!("before: {:?}", PoolAssets::owner(lp_token));

			asset_hub_westend_runtime::migrations::NativeAssetParents0ToParents1Migration::<Runtime>::on_runtime_upgrade();

			assert_eq!(0, Balances::balance(&old_pool_account));
			assert_ne!(0, Balances::balance(&new_pool_account));
			assert_eq!(0, Assets::balance(ASSET_ID, old_pool_account.clone()));
			assert_ne!(0, Assets::balance(ASSET_ID, new_pool_account.clone()));
			assert_eq!(0, PoolAssets::balance(lp_token, old_pool_account.clone()));
			assert_ne!(0, PoolAssets::balance(lp_token, new_pool_account.clone()));
			assert_eq!(Some(new_pool_account), PoolAssets::owner(lp_token));
			println!("after: {:?}", PoolAssets::owner(lp_token));
		})
}
```


but it only works with this hardcoded hack, so I better not pushed it
```
	fn is_native(asset_id: &Box<MultiLocation>) -> bool {
		*asset_id == Self::get_native() ||
                        // THIS IS HACK FOR MIGRATION
			*asset_id == Box::new(MultiLocation::new(0, xcm::latest::prelude::Here))
	}

```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3217962

**Comment by bkontur** [comment]
> > Looks OK, @jsidorenko can check for the migration especially?
> 
> for that migration, I have a working test:
> 
> ```
> pub use asset_hub_westend_runtime::{AssetConversion, PoolAssets};
> 
> #[test]
> fn test_migration() {
> 	use frame_support::traits::{
> 		fungible::Inspect, fungibles::roles::Inspect as InspectRoles, OnRuntimeUpgrade,
> 	};
> 
> 	let account = AccountId::from(ALICE);
> 	let existential_deposit = ExistentialDeposit::get();
> 
> 	ExtBuilder::<Runtime>::default()
> 		.with_collators(vec![AccountId::from(ALICE)])
> 		.with_session_keys(vec![(
> 			AccountId::from(ALICE),
> 			AccountId::from(ALICE),
> 			SessionKeys { aura: AuraId::from(sp_core::sr25519::Public::from_raw(ALICE)) },
> 		)])
> 		.with_balances(vec![
> 			(account.clone(), existential_deposit * AssetDeposit::get() * 100),
> 		])
> 		.build()
> 		.execute_with(|| {
> 			const ASSET_ID: u32 = 1;
> 			let asset_native = Box::new(MultiLocation::new(0, Here));
> 			let asset_one = Box::new(MultiLocation {
> 				parents: 0,
> 				interior: X2(PalletInstance(50), GeneralIndex(ASSET_ID.into())),
> 			});
> 			assert_ok!(Assets::create(
> 				RuntimeHelper::<Runtime>::origin_of(account.clone()),
> 				ASSET_ID.into(),
> 				account.clone().into(),
> 				1000,
> 			));
> 			assert_ok!(Assets::mint(
> 				RuntimeHelper::<Runtime>::origin_of(account.clone()),
> 				ASSET_ID.into(),
> 				account.clone().into(),
> 				3_000_000_000_000,
> 			));
> 			assert_ok!(AssetConversion::create_pool(
> 				RuntimeHelper::<Runtime>::origin_of(account.clone()),
> 				asset_native.clone(),
> 				asset_one.clone(),
> 			));
> 			assert_ok!(AssetConversion::add_liquidity(
> 				RuntimeHelper::<Runtime>::origin_of(account.clone()),
> 				asset_native.clone(),
> 				asset_one.clone(),
> 				1_000_000_000_000,
> 				2_000_000_000_000,
> 				0,
> 				0,
> 				account.into()
> 			));
> 
> 			let lp_token = PoolAssets::asset_ids().next().unwrap();
> 
> 			let old_pool_id = AssetConversion::get_pool_id(asset_native.clone(), asset_one.clone());
> 			let old_pool_account = AssetConversion::get_pool_account(&old_pool_id);
> 			let new_pool_id = AssetConversion::get_pool_id(Box::new(WestendLocation::get()), asset_one.clone());
> 			let new_pool_account = AssetConversion::get_pool_account(&new_pool_id);
> 
> 			assert_ne!(0, Balances::balance(&old_pool_account));
> 			assert_eq!(0, Balances::balance(&new_pool_account));
> 			assert_ne!(0, Assets::balance(ASSET_ID, old_pool_account.clone()));
> 			assert_eq!(0, Assets::balance(ASSET_ID, new_pool_account.clone()));
> 			assert_ne!(0, PoolAssets::balance(lp_token, old_pool_account.clone()));
> 			assert_eq!(0, PoolAssets::balance(lp_token, new_pool_account.clone()));
> 			assert_eq!(Some(old_pool_account.clone()), PoolAssets::owner(lp_token));
> 			println!("before: {:?}", PoolAssets::owner(lp_token));
> 
> 			asset_hub_westend_runtime::migrations::NativeAssetParents0ToParents1Migration::<Runtime>::on_runtime_upgrade();
> 
> 			assert_eq!(0, Balances::balance(&old_pool_account));
> 			assert_ne!(0, Balances::balance(&new_pool_account));
> 			assert_eq!(0, Assets::balance(ASSET_ID, old_pool_account.clone()));
> 			assert_ne!(0, Assets::balance(ASSET_ID, new_pool_account.clone()));
> 			assert_eq!(0, PoolAssets::balance(lp_token, old_pool_account.clone()));
> 			assert_ne!(0, PoolAssets::balance(lp_token, new_pool_account.clone()));
> 			assert_eq!(Some(new_pool_account), PoolAssets::owner(lp_token));
> 			println!("after: {:?}", PoolAssets::owner(lp_token));
> 		})
> }
> ```
> 
> but it only works with this hardcoded hack, so I better not pushed it
> 
> ```
> 	fn is_native(asset_id: &Box<MultiLocation>) -> bool {
> 		*asset_id == Self::get_native() ||
>                         // THIS IS HACK FOR MIGRATION
> 			*asset_id == Box::new(MultiLocation::new(0, xcm::latest::prelude::Here))
> 	}
> ```

actually, the migration misses the real pool_id change, omg... I am on it

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3218047

File: parachains/runtimes/assets/asset-hub-westend/src/lib.rs [hunk 4]
```diff
@@ -1360,3 +1376,120 @@ cumulus_pallet_parachain_system::register_validate_block! {
...
+					pool_asset_id.clone(),
+					&old_pool_account,
+					&new_pool_account,
+					T::PoolAssets::balance(pool_asset_id.clone(), &old_pool_account),
+					Preservation::Expendable,
+				);
+				reads.saturating_accrue(1);
+				writes.saturating_accrue(2);
+
+				// change owner ship of LP token
```
Comment (jsidorenko) [inline parachains/runtimes/assets/asset-hub-westend/src/lib.rs:1463]:
```suggestion
				// change the ownership of LP token
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3218700

**Review by gilescope — APPROVED** [review]

**Comment by bkontur** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3235604

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/3237185

**Review by bkchr — APPROVED** [review]

**Pull request merged**